### PR TITLE
mgr: centralized setting/getting of mgr configs  

### PR DESCRIPTION
**Pull request by Rubab-Syed**

### CODE REVIEW

File: src/mon/ConfigMonitor.cc [hunk 1]
```diff
@@ -529,26 +531,18 @@ void ConfigMonitor::load_config()
...
-    }
-
     string section_name;
-    MaskedOption mopt(opt);
-    mopt.raw_value = value;
-    if (who.size() &&
-	!ConfigMap::parse_mask(who, &section_name, &mopt.mask)) {
-      derr << __func__ << " ignoring key " << key << dendl;
-    } else {
+    if (key.find("mgr") != std::string::npos) {
```
Comment (liewegas) [inline src/mon/ConfigMonitor.cc]: if (key.find("mgr") == 0) ?

File: src/mon/ConfigMonitor.cc [hunk 2]
```diff
@@ -529,26 +531,18 @@ void ConfigMonitor::load_config()
...
+      string opt_name = key.substr(key.find('/') + 1);       // opt_name = mgr/dashboard/server_port
+      if (opt_name.substr(0, 3) == "mgr") {
+	name = opt_name.substr(4);                    // name = dashboard/server_port
+      } else {
+	name = opt_name;
```
Comment (liewegas) [inline src/mon/ConfigMonitor.cc]: I don't follow what this case is for...

**Comment by liewegas** [comment]
This looks reasonable!  I think we should plan to eventually have hte mgr modules advertise a schema for their options (names, help strings, types, level, all that stuff) but this will get us started.

File: src/mgr/ActivePyModules.cc [hunk 3]
```diff
@@ -495,6 +514,47 @@ void ActivePyModules::set_config(const std::string &module_name,
...
+
+    std::ostringstream cmd_json;
+    JSONFormatter jf;
+    jf.open_object_section("cmd");
+    if (val) {
+      jf.dump_string("prefix", "config set mgr");
+      jf.dump_string("key", global_key);
+      jf.dump_string("val", *val);
+    } else {
+      jf.dump_string("prefix", "config-key del");
```
Comment (jcsp) [inline src/mgr/ActivePyModules.cc]: This should be converted from `config-key del` to `config rm`

File: src/mgr/PyModuleRegistry.cc [hunk 4]
```diff
@@ -116,7 +116,8 @@ void PyModuleRegistry::standby_start(MonClient *monc)
 
   dout(4) << "Starting modules in standby mode" << dendl;
 
-  standby_modules.reset(new StandbyPyModules(monc, mgr_map, clog));
+  standby_modules.reset(new StandbyPyModules(
```
Comment (jcsp) [inline src/mgr/PyModuleRegistry.cc]: Don't forget to make the equivalent changes to ActivePyModules initialization to pass the module_config in there too

File: src/mgr/PyModuleRegistry.h [hunk 5]
```diff
@@ -40,6 +40,7 @@ class PyModuleRegistry
 {
 private:
   mutable Mutex lock{"PyModuleRegistry::lock"};
+  mutable Mutex config_lock{"PyModuleRegistry::config_lock"};
```
Comment (jcsp) [inline src/mgr/PyModuleRegistry.h]: A reference to this lock will be needed by anyone who has a reference to `module_config` (e.g. StandbyPymodules and soon ActivePyModules), it's probably easiest to do that by bundling the Mutex and std::map together into a little class.

**Comment by jcsp** [comment]
I've done some work on top of this and opened https://github.com/ceph/ceph/pull/21442, which now supersedes this PR

**Pull request closed without merging**